src/ring_theory/derivation.lean

```diff
@@ -0,0 +1,223 @@
+
+@[ext] theorem ext (H : ∀ a, D1 a = D2 a) : D1 = D2 :=
+coe_injective $ funext H
+
+@[simp] lemma map_add : D (a + b) = D a + D b := is_add_hom.map_add D a b
+@[simp] lemma map_zero : D 0 = 0 := is_add_monoid_hom.map_zero D
+@[simp] lemma map_smul : D (r • a) = r • D a := linear_map.map_smul D r a
+@[simp] lemma leibniz : D (a * b) = a • D b + b • D a := D.leibniz' _ _
+
+@[simp] lemma map_one_eq_zero : D 1 = 0 :=
+begin
+  have h : D 1 = D (1 * 1) := by rw mul_one,
+  rw [leibniz D 1 1, one_smul] at h,
+  exact zero_left_cancel h,
+end
+
+@[simp] lemma map_algebra_map : D (algebra_map R A r) = 0 :=
+by rw [←mul_one r, ring_hom.map_mul, map_one, ←smul_def, map_smul, map_one_eq_zero, smul_zero]
+
+instance : has_zero (derivation R A M) :=
+⟨⟨(0 : A →ₗ[R] M), λ a b, by simp only [add_zero, linear_map.zero_apply,
+                                      linear_map.to_fun_eq_coe, smul_zero]⟩⟩
+
+instance : inhabited (derivation R A M) := ⟨0⟩
+
+instance : add_comm_monoid (derivation R A M) :=
+{ add := λ D1 D2, ⟨D1 + D2, λ a b, begin
+  simp only [leibniz, linear_map.add_apply, linear_map.to_fun_eq_coe, coe_linear_map, smul_add],
+  cc, end⟩,
+  add_assoc := λ D E F, ext $ λ a, add_assoc _ _ _,
+  zero_add := λ D, ext $ λ a, zero_add _,
+  add_zero := λ D, ext $ λ a, add_zero _,
+  add_comm := λ D E, ext $ λ a, add_comm _ _,
+  ..derivation.has_zero }
+
+@[simp] lemma add_apply : (D1 + D2) a = D1 a + D2 a := rfl
+
+@[priority 100]
+instance derivation.Rsemimodule : semimodule R (derivation R A M) :=
+{ smul := λ r D, ⟨r • D, λ a b, by simp only [linear_map.smul_apply, leibniz,
+  linear_map.to_fun_eq_coe, algebra_compatible_smul_comm, coe_linear_map, smul_add, add_comm],⟩,
+  mul_smul := λ a1 a2 D, ext $ λ b, mul_smul _ _ _,
+  one_smul := λ D, ext $ λ b, one_smul _ _,
+  smul_add := λ a D1 D2, ext $ λ b, smul_add _ _ _,
+  smul_zero := λ a, ext $ λ b, smul_zero _,
+  add_smul := λ a1 a2 D, ext $ λ b, add_smul _ _ _,
+  zero_smul := λ D, ext $ λ b, zero_smul _ _ }
+
+@[simp] lemma smul_to_linear_map_coe : ↑(r • D) = (r • D : A →ₗ[R] M) := rfl
+@[simp] lemma smul_apply : (r • D) a = r • D a := rfl
+
+instance : semimodule A (derivation R A M) :=
+{ smul := λ a D, ⟨⟨λ b, a • D b,
+    λ a1 a2, by rw [D.map_add, smul_add],
+    λ a1 a2, by rw [D.map_smul, algebra_compatible_smul_comm]⟩,
+    λ b c, by { dsimp, simp only [smul_add, leibniz, smul_comm, add_comm] }⟩,
+  mul_smul := λ a1 a2 D, ext $ λ b, mul_smul _ _ _,
+  one_smul := λ D, ext $ λ b, one_smul A _,
+  smul_add := λ a D1 D2, ext $ λ b, smul_add _ _ _,
+  smul_zero := λ a, ext $ λ b, smul_zero _,
+  add_smul := λ a1 a2 D, ext $ λ b, add_smul _ _ _,
+  zero_smul := λ D, ext $ λ b, zero_smul A _ }
+
+/-- The composition of a derivation and a linear map is a derivation. -/
+def comp {N : Type*} [add_cancel_comm_monoid N] [semimodule A N] [semimodule R N]
+  [is_scalar_tower R A N]
+  (D : derivation R A M) (f : M →ₗ[A] N) : derivation R A N :=
```

Comment:
It would be clearer if the order of `f` and `D` in `comp f D` match the order in `comp f D a = f (D a)`.
```suggestion
  (f : M →ₗ[A] N) (D : derivation R A M) : derivation R A N :=
```

Reply:
I believe I copied this lemma from Kenny's last year implementation, and I guess he chose this order so to have `D.comp f`. Is there no other naming that could save this notation? Maybe putting comp in linear_map namespace?

Reply:
The notation would still work. Dot notation takes the types into account. (It will use `D` as the first argument of type `derivation`.)

Reply:
It'll work with `D.comp f` but not `f.comp D`. In order to fully apply Anne's suggestion I might prefer to define `comp_der` in the `linear_map` namespace.